components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/DefaultSAML2SSOManager.java

```diff
@@ -833,29 +906,29 @@ private XMLObject unmarshall(String samlString) throws SAMLSSOException {
             Element element = document.getDocumentElement();
             UnmarshallerFactory unmarshallerFactory = Configuration.getUnmarshallerFactory();
             Unmarshaller unmarshaller = unmarshallerFactory.getUnmarshaller(element);
-            response = unmarshaller.unmarshall(element);
-
-            // Checking for duplicate samlp:Response. This is done to thwart possible XSW attacks
-            NodeList responseList = response.getDOM().getElementsByTagNameNS(SAMLConstants.SAML20P_NS, "Response");
-            if (responseList.getLength() > 0) {
-                log.error("Invalid schema for the SAML2 response. Multiple Response elements found.");
-                throw new SAMLSSOException("Error occurred while processing SAML2 response.");
-            }
-
-            // Checking for multiple Assertions. This is done to thwart possible XSW attacks.
-            NodeList assertionList = response.getDOM().getElementsByTagNameNS(SAMLConstants.SAML20_NS, "Assertion");
-            if (assertionList.getLength() > 1) {
-                log.error("Invalid schema for the SAML2 response. Multiple Assertion elements found.");
-                throw new SAMLSSOException("Error occurred while processing SAML2 response.");
-            }
-
-            return response;
+            return unmarshaller.unmarshall(element);
         } catch (ParserConfigurationException | UnmarshallingException | SAXException | IOException e) {
             throw new SAMLSSOException("Error in unmarshalling SAML Request from the encoded String", e);
         }
 
     }
 
+    private void validateResponseFormat(XMLObject response) throws SAMLSSOException {
+        // Checking for duplicate samlp:Response. This is done to thwart possible XSW attacks
+        NodeList responseList = response.getDOM().getElementsByTagNameNS(SAMLConstants.SAML20P_NS, "Response");
+        if (responseList != null && responseList.getLength() > 0) {
+            log.error("Invalid schema for the SAML2 response. Multiple Response elements found.");
+            throw new SAMLSSOException("Error occurred while processing SAML2 response.");
+        }
+
+        // Checking for multiple Assertions. This is done to thwart possible XSW attacks.
+        NodeList assertionList = response.getDOM().getElementsByTagNameNS(SAMLConstants.SAML20_NS, "Assertion");
+        if (assertionList != null && assertionList.getLength() > 1) {
+            log.error("Invalid schema for the SAML2 response. Multiple Assertion elements found.");
```

Comment:
If you log it and rethrow it, then there's no way for the upstream code to know that you've already logged the exception, and so the same exception might get logged twice.